### PR TITLE
feat: add an autofix for inserting an unsafe block to missing unsafe diagnostic

### LINKED ISSUE #14241: missing unsafe diagnostic should have an autofix for inserting an unsafe block
Relevant code is here https://github.com/rust-lang/rust-analyzer/blob/master/crates/ide-diagnostics/src/handlers/missing_unsafe.rs

### PR DESCRIPTION
**Pull request by XFFXFF**
close #14241

### CODE REVIEW

**Comment by Veykril** [comment]
Ah sorry, I mislabeled the issue with assist. This should be the quickfix for the diagnostic that was linked in the issue here https://github.com/rust-lang/rust-analyzer/blob/master/crates/ide-diagnostics/src/handlers/missing_unsafe.rs

For an example diagnostic that has a quickfix to see how that's done you can take a look at https://github.com/rust-lang/rust-analyzer/blob/master/crates/ide-diagnostics/src/handlers/missing_fields.rs

(fortunately assists and quickfixes don't differ too much)

**Comment by XFFXFF** [comment]
@Veykril Don't worry about it, the process of implementation let me understand RA better

**Comment by XFFXFF** [comment]
It is much easier than implementing assist, mainly because I don't have to determine which ones are unsafe

File: crates/ide-diagnostics/src/handlers/missing_unsafe.rs [hunk 1]
```diff
@@ -9,11 +16,60 @@ pub(crate) fn missing_unsafe(ctx: &DiagnosticsContext<'_>, d: &hir::MissingUnsaf
...
+// ```
+// We can't add an unsafe block like this:
+// ```
+// let z = unsafe { STATIC_MUT } .a;
+// ```
+fn pick_best_node_to_add_unsafe_block(
+    ctx: &DiagnosticsContext<'_>,
+    expr: &ast::Expr,
+) -> SyntaxNode {
+    let Some(let_or_expr_stmt) = ctx.sema.ancestors_with_macros(expr.syntax().clone()).find(|node| {
```
Comment (Veykril) [inline crates/ide-diagnostics/src/handlers/missing_unsafe.rs:61]: We need to be sure not to suggest fixes for when this diagnostic comes from a macro, so we should instead do something like
```rs
ctx.sema.original_ast_node(expr).syntax().ancestors()...
```

File: crates/ide-diagnostics/src/handlers/missing_unsafe.rs [hunk 2]
```diff
@@ -9,11 +16,60 @@ pub(crate) fn missing_unsafe(ctx: &DiagnosticsContext<'_>, d: &hir::MissingUnsaf
...
+        LetStmt::can_cast(node.kind()) || ExprStmt::can_cast(node.kind())
+    }) else {
+        // Is this reachable?
```
Comment (Veykril) [inline crates/ide-diagnostics/src/handlers/missing_unsafe.rs:64]: Yes it it is, `const Foo: () = expr;` here expr is not a statement

File: crates/ide-diagnostics/src/handlers/missing_unsafe.rs [hunk 3]
```diff
@@ -9,11 +16,60 @@ pub(crate) fn missing_unsafe(ctx: &DiagnosticsContext<'_>, d: &hir::MissingUnsaf
...
+//
+// Or this example:
+// ```
+// let z = STATIC_MUT.a;
+// ```
+// We can't add an unsafe block like this:
+// ```
+// let z = unsafe { STATIC_MUT } .a;
+// ```
+fn pick_best_node_to_add_unsafe_block(
```
Comment (Veykril) [inline crates/ide-diagnostics/src/handlers/missing_unsafe.rs:57]: I think we can do better than wrapping a statement here, (also wrapping a let statement will break things as the`unsafe {..}` introduces a new scope).

So I'd say let's pick the first ancestor expression of the unsafe `expr` that is not a receiver to a method call or field access, the left hand side of an assignment expression or the target of a reference expression. As all of those cases would incur a forced move if wrapped which might not be wanted. That is:

`unsafe_expr.foo` -> `unsafe { unsafe_expr.foo }`
`unsafe_expr.foo.bar` -> `unsafe { unsafe_expr.foo.bar }`
`unsafe_expr.foo()` -> `unsafe { unsafe_expr.foo() }`
`unsafe_expr.foo.bar()` -> `unsafe { unsafe_expr.foo.bar() }`
`unsafe_expr += something` -> `unsafe { unsafe_expr += something }`
`&unsafe_expr` -> `unsafe { &unsafe_expr }`

but anything else can just be wrapped, few examples:

`foo(unsafe_expr)` -> `foo(unsafe { unsafe_expr })`
`let x = unsafe_expr` -> `let x = unsafe { unsafe_expr }`

Comment (XFFXFF) [reply crates/ide-diagnostics/src/handlers/missing_unsafe.rs:57]: Thank you for your suggestion. That’s a great idea! I will try it out

File: crates/ide-diagnostics/src/handlers/missing_unsafe.rs [hunk 4]
```diff
@@ -9,11 +15,99 @@ pub(crate) fn missing_unsafe(ctx: &DiagnosticsContext<'_>, d: &hir::MissingUnsaf
...
+                    // text range of the `node` (which is the ancestor of the
+                    // `unsafe_expr`) is contained in the text range of the
+                    // receiver. The `node` could potentially be the receiver, the
+                    // method name, or the argument list. Since the `node` is the
+                    // ancestor of the unsafe_expr, it cannot be the method name.
+                    // Additionally, if the `node` is the argument list, the loop
+                    // would break at least when `parent` reaches the argument list.
+                    //
+                    // Dispite this, I still check the text range because I think it
+                    // makes the code easier to understand.
```
Comment (XFFXFF) [inline crates/ide-diagnostics/src/handlers/missing_unsafe.rs:76]: How do you see it? @Veykril

File: crates/ide-diagnostics/src/handlers/missing_unsafe.rs [hunk 5]
```diff
@@ -9,11 +15,99 @@ pub(crate) fn missing_unsafe(ctx: &DiagnosticsContext<'_>, d: &hir::MissingUnsaf
         "this operation is unsafe and requires an unsafe function or block",
         ctx.sema.diagnostics_display_range(d.expr.clone().map(|it| it.into())).range,
     )
+    .with_fixes(fixes(ctx, d))
+}
+
+fn fixes(ctx: &DiagnosticsContext<'_>, d: &hir::MissingUnsafe) -> Option<Vec<Assist>> {
+    let root = ctx.sema.db.parse_or_expand(d.expr.file_id)?;
```
Comment (Veykril) [inline crates/ide-diagnostics/src/handlers/missing_unsafe.rs:22]:
```suggestion
    if d.expr.file_id.is_macro() {
        return None;
    }
    let root = ctx.sema.db.parse_or_expand(d.expr.file_id)?;
```
The fixit will not work correctly for macro expansions, so for the time being let's disable this until we have a decent way of handling these cases in general.

**Comment by XFFXFF** [comment]
@Veykril I have modified the code according to your suggestion, the changes can be found at https://github.com/rust-lang/rust-analyzer/pull/14281/commits/0c935732bcb7f13a491181e8996be3ec1c2244ac, please take a look again

**Comment by XFFXFF** [comment]
I don't know why [CI/TypeScript (windows-latest)](https://github.com/rust-lang/rust-analyzer/actions/runs/4421728958/jobs/7752831744) didn't pass

**Comment by lnicola** [comment]
https://user-images.githubusercontent.com/308347/226258052-07f7c722-bdc2-49bb-b0af-dac7877c761f.mp4

**Pull request merged**